walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,70 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.Collections;
+import java.util.ArrayList;
+import java.util.List;
 
 public final class FindMeetingQuery {
+  
+  public final Collection<TimeRange> slots = new ArrayList<>();
+  public final List<TimeRange> bookedTimes = new ArrayList<>();
+
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+
+    if(request.getDuration() >= TimeRange.WHOLE_DAY.duration()){
+        return slots;
+    }else if(events.isEmpty()){
+        slots.add(TimeRange.WHOLE_DAY);
+    }else {
+        for(Event event: events){
+            for(String attendee: request.getAttendees()){
+                if(event.getAttendees().contains(attendee)){
+                    bookedTimes.add(event.getWhen());
+                    break;
+                }
+            }
+        }
+
+        if(bookedTimes.size() == 0){
+            slots.add(TimeRange.WHOLE_DAY);
+            return slots;
+        }
+
+        //Sort the booked times by start of day
+        Collections.sort(bookedTimes, TimeRange.ORDER_BY_START);
+
+        //Merge overlapping or nested timeslots into a single chunk of time
+        for (int i = 0; i < bookedTimes.size() - 1; i++){
+            if(bookedTimes.get(i).overlaps(bookedTimes.get(i+1)) || bookedTimes.get(i).contains(bookedTimes.get(i+1))){
+                if(bookedTimes.get(i).end() > bookedTimes.get(i+1).end()){
+                    bookedTimes.set(i, TimeRange.fromStartEnd(bookedTimes.get(i).start(), bookedTimes.get(i).end(), false));
+                }else{
+                    bookedTimes.set(i, TimeRange.fromStartEnd(bookedTimes.get(i).start(), bookedTimes.get(i+1).end(), false));
+                }
+                bookedTimes.remove(i+1);
```

Comment:
won't this affect the for loop condition?

Reply:
I don't think it would. The for loop runs directly upto bookedTimes.size() - 1 (the size is not stored in a variable that is being compared to). Deleting an entry would also reduce the size of the list. 
Please let me know what you think

Reply:
Yes! It still works as expected and all my test cases pass.

Reply:
Let's say bookedTimes is [(1,3), (2,8), (4,6)] before you enter the loop. If I understand correctly, the expected output is [(1,8)] after the loop. But your loop will return [(1,8), (4,6)].

Reply:
Oooh. I see. 
I think a solution to this would be to run a while loop upto size() - 1, and only increment [i] if we 
don't delete the next entry, 
So something like this:

int i = 0;
while (i < bookedTimes.size() - 1){
            if(bookedTimes[i] and bookedTimes[i+1] overlap){
                // endTime = max(bookedTimes.get(i).end(), bookedTimes.get(i+1).end())
                set bookedTimes[i] to TimeRange upto bookedTimes[i+1}.end();
                bookedTimes.remove(i+1);
            }else{
                i++;
            }
        }


This would then breakdown the array as this:
0th iteration: [(1,3), (2,8), (4,6)] // i = 0; length - 1 = 2
1st iteration: [(1,8), (4,6)]    // i = 0; length - 1 = 1
2nd iteration: [(1,8)] // i = 1; length - 1 = 1
//breaks

Reply:
Yes that would work. I'm surprised none of the existing tests caught this though.

Reply:
Yep. Wouldn't have caught that either. They only give two to three time ranges in the tests. So I don't think an overlap like that was tested